add tests for Retry annotation retrieval

McFoggy opened the pull request:
add tests to check the correctness of the annotations picked up when
evaluating @Retry annotation values.
tests include: class inheritance, method override, annotation removal.

Emily-Jiang commented:
Thanks @McFoggy for your contribution! Can you create an issue for this PR please?

McFoggy commented:
There is one already existing #183 that is referenced in the commit.

neilgsyoung commented:
These are really useful additional tests. They need a couple of minor changes to run successfully:

1/  The tests need to call the appropriate RetryService in RetryVisibilityTest. They currently all call the baseService.

`    @Test
    public void serviceOverrideClassLevelUsesClassLevelAnnotation() {
        int nbExpectedRetries = 4;  // see RetryServiceOverrideClassLevel class annotation

        checkServiceCall(nbExpectedRetries, this.serviceOverrideClassLevel, // <-replace baseService 
"serviceOverrideClassLevelUsesClassLevelAnnotation");
    }

    @Test
    public void serviceOverrideMethodLevelUsesMethodLevelAnnotation() {
        int nbExpectedRetries = 4;   // see RetryServiceOverrideMethodLevel#service() method annotation

        checkServiceCall(nbExpectedRetries, this.serviceOverrideMethodLevel,  // <-replace baseService
"serviceOverrideMethodLevelUsesMethodLevelAnnotation");
    }
    
    @Test
    public void serviceRetryRemovedAtMethodLevel() {
        int nbExpectedRetries = 0;  // see RetryServiceMethodSuppressLevel#service() method with no annotation

        checkServiceCall(nbExpectedRetries, this.serviceSuppressMethodLevel,  // <-replace baseService
"serviceRetryRemovedAtMethodLevel");
    }`

2/ The RetryServiceMethodSuppressLevel class needs reworking. It will currently inherit the default Retry policy from the BaseRetryService and the service will be executed 4 times. 
One way to rework this is to...
`public class RetryServiceMethodSuppressLevel extends AlternateRetryServiceOverrideMethodLevel`

....where AlternateRetryServiceOverrideMethodLevel, looks like this...

`@RequestScoped
@RS(RetryServiceType.METHOD_OVERRIDE)
public class AlternateRetryServiceOverrideMethodLevel extends AlternateBaseRetryService {
    @Override
    @Retry(maxRetries = 4)
    public Connection service() throws IOException {
        return super.service();
    }
}`

and AlternateBaseRetryService looks like this...

`@RequestScoped
@RS(RetryServiceType.BASE)
public class AlternateBaseRetryService implements RetryService {
    private int nbCalls = 0;

    @Override
    public Connection service() throws IOException {
        nbCalls++;
        return delegate();
    }

    protected Connection delegate() throws IOException {
        throw new IOException("cannot access delegate service");
    }

    @Override
    public int getNumberOfServiceCalls() {
        return nbCalls;
    }
}`

McFoggy commented:
@neilgsyoung can you please correct the formatting of your comment?

neilgsyoung commented:
That hopefully looks better.

johnament commented on tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/visibility/retry/RetryServiceMethodSuppressLevel.java:
```diff
@@ -0,0 +1,37 @@
+
+@RequestScoped
+@RS(RetryServiceType.METHOD_SUPPRESS)
+public class RetryServiceMethodSuppressLevel extends BaseRetryService {
```
Don't forget that since the FT annotations are all `@Inherited` the original annotations will be retained.  E.g. calling `RetryServiceMethodSuppressLevel.class.getMethod("service").getAnnotation(Retry.class)` will still return the object.

I don't think your test for this will pass for anyone.

McFoggy replied:
@johnament this came from #183 and the answer of from @Emily-Jiang to your issue.

I have no pb in keeping it or removing it

> I don't think your test for this will pass for anyone.

unless implementations use `getDeclaredAnnotation()` instead of `getAnnotation()`

Honestly, I see interests in both way of handling this annotation retrieval.
It is a matter of decision & balance between the strict conformity & the facility offer to disable a feature by overriding a method.
Moreover, looking at the javadoc of `@Inherited`, no strict description is done when it comes to method annotation lookup.

If there were votes available on an issue in github we could go with a democratic way ; instead I'd expect project leads to take a decision.

johnament replied:
Well, the project leads would be John C or Kevin.  I'm not in agreement with Emily's POV, since these are `@Inherited`.  I would recommend that we remove `@Inherited` if this is the behavior expected.

Emily-Jiang replied:
Since the subclass inherits the annotations, the method will have the annotations in RetryServiceMethodSuppressLevel. Actually, think it further. It might be better not to make them inherited, because we cannot easily get rid of the annotations. If they are not inherited, we can always add them. We can discuss this on tomorrow's (15th Nov) call as well.

johnament replied:
Isn't that what I had just recommended, removing `@Inherited`?

Emily-Jiang replied:
Yes, John. I think I am not in disagreement with you on this. Since this is behaviour changes, we should have further discussion on tomorrow's call.

McFoggy commented:
@neilgsyoung let's wait for the final conclusions on how to handle the annotations to take into account your second remark (or a reworked version).

also 
- I added a simple test that  demo what happens at J2SE level for our `Retry` annotation retrieval
- I prepared also new scenarios to be tested

McFoggy commented:
FYI I started a boring work which goal is to understand existing behaviors of annotation retrieval in established & already specified areas (EJBs & JTA). 

You'll find the project there: https://github.com/McFoggy/jee-tx-resolution.

I'll continue to enhance it to add some missing coverage and to run it on different containers ; but honestly it already surprised me in how EJB spec and JTA spec were behaving for some corner cases (at least under wildfly).  
It will deserve to have an overview (in a table probably) of the demoed behaviors and a README.

McFoggy commented:
@antoinesd as discussed today I added missing tests.

I have rebased and will squash on demand.

My impl does not pass the tests because currently I was looking for the annotations by hand first looking for annotation on method then on class and redo the same process on parent classes.  
From the understanding I had of the discussions on inheritance you had with @Emily-Jiang, I implemented the tests on the TCK in this PR more to enable a pure delegation to `java.lang.reflect.Method.getAnnotation()` & `java.lang.class#getAnnotation()`

Emily-Jiang commented:
@McFoggy we hit some issues when running the tests. Did you fix the issue @neilgsyoung suggested?

McFoggy commented:
> Did you fix the issue @neilgsyoung suggested?

@Emily-Jiang yes I think I did, but I can have make mistakes.

> we hit some issues when running the tests.

The tests are there to express an order of retrieval after the discussions that occured. 
In my opinion the question is not that Liberty or any other implementation does or does not pass the tests, the question is : are the tests reflecting the retrieval pattern that is expected in the spec.

McFoggy commented:
I am even not sure about the expectations I have put in the tests.  
Let's take the example of `RetryVisibilityTest#serviceOverrideClassLevelUsesClassLevelAnnotation`.  
We have a common service

````java
@Retry            // use defaults maxRetries = 3
BaseService {
     void service() {...}
}
````
which expresses: _define a retry policy of 3 for all exposed methods of this service and of sub implementations_ 

Then we have a specialization:
````java
@Retry(maxRetries = 4)
Specialization1Service extends BaseService {
}
````
which I assume means: _for the Specialization1 use a retry policy of 4_

Thus comes the question:
- do we expect the number of retry to be really 4 or 3 when calling _`service()`_ method on an instance of `Specialization1Service`.

If the implementation uses the annotation found starting from the method location, then obviously it will be 3 ; is that expected? If the implementation uses first the target object of the call then it will be 4.
The tests are written to expose coverage of such cases. 
If we think it is 3 then let's put 3 in the test expectation, if we think it is 4 then let's put 4 in the test expectation. The PR provided the test coverage, let's review it now to discuss the expectations.  
They are not always easy to define...

Emily-Jiang commented:
> Thus comes the question:

    do we expect the number of retry to be really 4 or 3 when calling service() method on an instance of Specialization1Service.

If the implementation uses the annotation found starting from the method location, then obviously it will be 3 ; is that expected? If the implementation uses first the target object of the call then it will be 4.
The tests are written to expose coverage of such cases.
If we think it is 3 then let's put 3 in the test expectation, if we think it is 4 then let's put 4 in the test expectation. The PR provided the test coverage, let's review it now to discuss the expectations.
They are not always easy to define...

From my understanding, it should be 4.

McFoggy commented:
> From my understanding, it should be 4.

the tests are written with this philosophy in mind or at least it was the intent...

McFoggy commented:
@antoinesd @Emily-Jiang @johnament @neilgsyoung any comment on the expectations of the tests?

neilgsyoung commented:
I need to investigate further but my initial run of the new visibility tests against OpenLiberty shows 15 passing and 5 failing tests...

Results :

Failed tests:
  RetryVisibilityTest>Arquillian.run:138->serviceBaseROCMNoRedefinition:215->checkServiceCall:248 in RetryVisibilityTest#serviceBaseROCMNoRedefinition service() should have been called exactly 4 times expected [4] but found [5]
  RetryVisibilityTest>Arquillian.run:138->serviceBaseROCMOverridedClassLevelNoMethodOverride:222->checkServiceCall:248 in RetryVisibilityTest#serviceBaseROCMOverridedClassLevelNoMethodOverride service() should have been called exactly 6 times expected [6] but found [5]
  RetryVisibilityTest>Arquillian.run:138->serviceBaseROMNoRedefinition:201->checkServiceCall:248 in RetryVisibilityTest#serviceBaseROMNoRedefinition service() should have been called exactly 1 times expected [1] but found [4]
  RetryVisibilityTest>Arquillian.run:138->serviceBaseROMOverridedClassLevelMethodOverride:187->checkServiceCall:248 in RetryVisibilityTest#serviceBaseROMOverridedClassLevelMethodOverride service() should have been called exactly 5 times expected [5] but found [4]
  RetryVisibilityTest>Arquillian.run:138->serviceRetryRemovedAtMethodLevel:159->checkServiceCall:248 in RetryVisibilityTest#serviceRetryRemovedAtMethodLevel service() should have been called exactly 1 times expected [1] but found [4]

Tests run: 20, Failures: 5, Errors: 0, Skipped: 0

Emily-Jiang commented:
@McFoggy can we go through these failing tests in the next hangout?

neilgsyoung commented:
Working through the failures that I see with OpenLiberty,

1. serviceBaseROCMNoRedefinition
Parent class: No annotation at class level, maxRetries=4 on method
Child class: No annotation at class level, no overriding method
TestExpectation: 3 retries
OpenLiberty TCK: 4 retries

Comment: The annotation on the service method in the parent class should be honoured, the service should be retried 4 times (5 executions)

2. serviceBaseROCMOverridedClassLevelNoMethodOverride
Parent class: No annotation at class level, maxRetries=4 on method
Child class: maxRetries=5 at class level, no overriding method
TestExpectation: 5 retries
OpenLiberty TCK: 4 retries

Comment: The annotation on the service method in the parent class should be honoured, the service should be retried 4 times (5 executions)

3. serviceBaseROMNoRedefinition
Parent class: No annotation at class level, default @Retry annotation on method
Child class: No annotation at class level, no overriding method
TestExpectation: 0 retries
OpenLiberty TCK: 3 retries

Comment: From org.eclipse.microprofile.faulttolerance.Retry, the default maxRetries value is 3,
the test should be changed to expect 4 executions of service().


4. serviceBaseROMOverridedClassLevelMethodOverride
Parent class: No annotation at class level, default @Retry annotation on method
Child class: maxRetries=4 at class level, no overriding method
TestExpectation: 4 retries
OpenLiberty TCK: 3 retries

Comment: The annotation on the service method in the parent class should be honoured.
From org.eclipse.microprofile.faulttolerance.Retry, the default maxRetries value is 3,
the test should be changed to expect 4 executions of service().

5. serviceRetryRemovedAtMethodLevel
Parent class: Default @Retry annotation at class level, no annotation on method
Child class: No annotation at class level, no annotation on overriding method
TestExpectation: 0 retries
OpenLiberty TCK: 3 retries

Comment: From org.eclipse.microprofile.faulttolerance.Retry, the default maxRetries value is 3,
the test should be changed to expect 4 executions of service().

neilgsyoung commented:
I'll use #213 to address the expected test results I described above.
